compiler/src/iree/compiler/Codegen/Dialect/GPU/Transforms/Transforms.cpp

```diff
@@ -287,6 +287,229 @@ LogicalResult fuseForallIntoConsumer(RewriterBase &rewriter,
+/// ```
+///   tensor.parallel_insert_slice ...
+///     tensor<1x2x3x4x5xf32> into tensor<?x?x3x?x5xf32>
+///   ...
+///   // Collapsed by:
+///   tensor.collapse_shape %result [[0, 1, 2], [3, 4]]
+///     tensor<?x?x3x?x5> into tensor<?x?x?xf32>
+/// ```
+/// Since the slice <1x2x3> is contiguous in <?x?x3> for the first group, and
+/// the slice <4x5> is contiguous in <?x5> for the second group.
+static LogicalResult
+collapsibleSlicePrecondition(RewriterBase &rewriter,
+                             tensor::ParallelInsertSliceOp sliceOp,
+                             SmallVector<ReassociationIndices> reassociations) {
+  if (!areAllConstantIntValue(sliceOp.getMixedStrides(), 1)) {
+    return rewriter.notifyMatchFailure(sliceOp, "strides are not all 1");
+  }
+  RankedTensorType sliceType = sliceOp.getSourceType();
+  if (sliceOp.getMixedSizes().size() != sliceType.getRank()) {
+    return rewriter.notifyMatchFailure(
+        sliceOp, "parallel insert slice is rank reducing");
+  }
+
+  SmallVector<OpFoldResult> sizes = sliceOp.getMixedSizes();
+  RankedTensorType fullTensorType = sliceOp.getDestType();
+  ArrayRef<int64_t> destShape = fullTensorType.getShape();
+  for (auto group : reassociations) {
+    bool isFullSlice = true;
+    for (auto idx : llvm::reverse(group)) {
+      std::optional<int64_t> constSize = getConstantIntValue(sizes[idx]);
+      // If the size is dynamic, then conservatively assume it is not full.
+      if (!constSize.has_value()) {
+        if (!isFullSlice) {
+          return rewriter.notifyMatchFailure(
+              sliceOp,
+              "parallel insert slice is not contiguous in the destination");
+        }
+        isFullSlice = false;
+        continue;
+      }
+      // Unit dimensions are always valid.
+      if (constSize.value() == 1) {
+        continue;
+      }
+      // If the size is not unit, then the slice must be full so far.
+      if (!isFullSlice) {
+        return rewriter.notifyMatchFailure(
+            sliceOp,
+            "parallel insert slice is not contiguous in the destination");
+      }
+      if (constSize.value() != destShape[idx]) {
+        isFullSlice = false;
+      }
+    }
+  }
+
+  return success();
+}
+
+/// Collapse all `ops` with the given `reassociations`. All `ops` are expected
+/// to have equivalent offsets, sizes, and strides. All strides are expected to
+/// be 1. This function assumes that the parallelInsertOp passes the
+/// collapsibleSlicePrecondition.
+static tensor::ParallelInsertSliceOp
+collapseParallelInsertOp(RewriterBase &rewriter,
+                         tensor::ParallelInsertSliceOp parallelInsertOp,
+                         SmallVector<ReassociationIndices> reassociations) {
+  // Compute the collapsed offsets, sizes, and strides.
+  rewriter.setInsertionPoint(parallelInsertOp.getParallelCombiningParent());
+  Location loc = parallelInsertOp.getParallelCombiningParent()->getLoc();
```

Comment:
I would just use the loc of the insert_slice here. The `forall.in_parallel` terminator isn't doing much of anything of note.